javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/ExtensionClassLoader.java

```diff
@@ -29,33 +38,83 @@
+
     // TODO add support for old deprecated property otel.javaagent.experimental.exporter.jar
-    List<URL> extension =
+    extensions.addAll(
         parseLocation(
             System.getProperty(
                 "otel.javaagent.experimental.extensions",
-                System.getenv("OTEL_JAVAAGENT_EXPERIMENTAL_EXTENSIONS")));
+                System.getenv("OTEL_JAVAAGENT_EXPERIMENTAL_EXTENSIONS")),
+            javaagentFile));
 
-    extension.addAll(
+    extensions.addAll(
         parseLocation(
             System.getProperty(
                 "otel.javaagent.experimental.initializer.jar",
-                System.getenv("OTEL_JAVAAGENT_EXPERIMENTAL_INITIALIZER_JAR"))));
+                System.getenv("OTEL_JAVAAGENT_EXPERIMENTAL_INITIALIZER_JAR")),
+            javaagentFile));
     // TODO when logging is configured add warning about deprecated property
 
-    if (extension.isEmpty()) {
+    if (extensions.isEmpty()) {
       return parent;
     }
 
-    List<ClassLoader> delegates = new ArrayList<>(extension.size());
-    for (URL url : extension) {
+    List<ClassLoader> delegates = new ArrayList<>(extensions.size());
+    for (URL url : extensions) {
       delegates.add(getDelegate(parent, url));
     }
     return new MultipleParentClassLoader(parent, delegates);
   }
 
-  private static List<URL> parseLocation(String locationName) {
+  private static void includeEmbeddedExtensionsIfFound(
+      ClassLoader parent, List<URL> extensions, File javaagentFile) {
+    try {
+      JarFile jarFile = new JarFile(javaagentFile, false);
+      Enumeration<JarEntry> entryEnumeration = jarFile.entries();
+      String prefix = "extensions/";
+      File tempDirectory = null;
+      while (entryEnumeration.hasMoreElements()) {
+        JarEntry jarEntry = entryEnumeration.nextElement();
+
+        if (jarEntry.getName().startsWith(prefix) && !jarEntry.isDirectory()) {
+          tempDirectory = ensureTempDirectoryExists(tempDirectory);
+
+          File tempFile = new File(tempDirectory, jarEntry.getName().substring(prefix.length()));
+          if (tempFile.createNewFile()) {
+            tempFile.deleteOnExit();
+            extractFile(jarFile, jarEntry, tempFile);
+            addFileUrl(extensions, tempFile);
+          } else {
+            System.out.println("Failed to create temp file " + tempFile);
+          }
+        }
+      }
+    } catch (IOException ex) {
+      System.err.println("Failed to open embedded extensions " + ex.getMessage());
+    }
+  }
+
+  private static File ensureTempDirectoryExists(File tempDirectory) throws IOException {
+    if (tempDirectory == null) {
+      tempDirectory = Files.createTempDirectory("otel-extensions").toFile();
+      tempDirectory.deleteOnExit();
+    }
+    return tempDirectory;
+  }
+
+  private static URLClassLoader getDelegate(ClassLoader parent, URL extensionUrl) {
+    return new URLClassLoader(new URL[] {extensionUrl}, parent);
```

Comment:
Shouldn't it return a new `ExtensionClassLoader`?
```suggestion
    return new ExtensionClassLoader(new URL[] {extensionUrl}, parent);
```

Reply:
oops :)